Calculate LineNode bounds, (Pen caps and Pen Thickness too)

danwalmsley opened the pull request:
## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
Correctly calculates the bounds of LineDraw calls.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
We dont correctly calculate the bounds and rendering artifacts occur.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
no artifacts

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->


## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
<!--- List any breaking changes here. When the PR is merged please add an entry to https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes -->


## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

Commit:
Add render demo showing the algoritm to calculate bounds for lines.

Commit:
dont inflate in draw operation

Commit:
move calculate bounds to LineNode.

Commit:
move to static LineBoundsHelper class.

Commit:
inflate rectangle nodes to take into account the pen thickness.

Commit:
correct the terminology of the helper methods

Commit:
fix draw operation tests.

Gillibald commented on src/Avalonia.Visuals/Rendering/SceneGraph/RectangleNode.cs:
```diff
@@ -30,7 +30,7 @@ internal class RectangleNode : BrushDrawOperation
             double radiusY,
             IDictionary<IVisual, Scene> childScenes = null)
-            : base(rect, transform, pen)
+            : base(rect.Inflate((pen?.Thickness ?? 0) / 2), transform, pen)
```
If two adjacent rects are drawn do the dirty rects align? If not is that the wanted behavior?

danwalmsley replied:
I dont know, @grokys probably has a better idea about what needs to be done here. this change is just maintaining the existing behavior.

danwalmsley replied:
@Gillibald the dirtyrects are snapped to pixels, and they also union if they partially overlap.

danwalmsley replied:
@grokys has resolved this.

Commit:
Merge branch 'master' into fixes/calculate-bounds-for-drawline

wieslawsoltes commented on src/Avalonia.Visuals/Rendering/SceneGraph/LineBoundsHelper.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public static Rect CalculateBounds(Point p1, Point p2, IPen p)
+        {
+            var angle = CalculateAngle(p1, p2);
```
I think doing all angle calculations in radians instead of converting to degrees would be better.

danwalmsley replied:
That would be at the limit of my maths skills!

wieslawsoltes replied:
Best to name variable `degrees` or `radians`.

danwalmsley replied:
done

jmacato commented on src/Avalonia.Visuals/Rendering/SceneGraph/LineBoundsHelper.cs:
```diff
@@ -0,0 +1,72 @@
+            var xDiff = p2.X - p1.X;
+            var yDiff = p2.Y - p1.Y;
+
+            return Math.Atan2(yDiff, xDiff) * 180.0 / Math.PI;
```
if my math skills hasnt rusted, Atan2 already returns radians, `180/Math.Pi` converts rads to degrees so if you remove this you can remove the `degreeToRadians` stuff in other functions

wieslawsoltes commented on src/Avalonia.Visuals/Rendering/SceneGraph/LineBoundsHelper.cs:
```diff
@@ -0,0 +1,72 @@
+            var xDiff = p2.X - p1.X;
+            var yDiff = p2.Y - p1.Y;
+
+            return Math.Atan2(yDiff, xDiff) * 180.0 / Math.PI;
```
Don't convert to degrees here:
```C#
return Math.Atan2(yDiff, xDiff);
```

danwalmsley replied:
done

wieslawsoltes commented on src/Avalonia.Visuals/Rendering/SceneGraph/LineBoundsHelper.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public static double CalculateOppSide(double angle, double hyp)
+        {
+            return Math.Sin(angle * degreeToRadians) * hyp;
```
Don't convert to radians here:
```C#
return Math.Sin(angle) * hyp;
```

danwalmsley replied:
done

wieslawsoltes commented on src/Avalonia.Visuals/Rendering/SceneGraph/LineBoundsHelper.cs:
```diff
@@ -0,0 +1,72 @@
+
+        public static double CalculateAdjSide(double angle, double hyp)
+        {
+            return Math.Cos(angle * degreeToRadians) * hyp;
```
Don't convert to radians here:
```C#
return Math.Cos(angle) * hyp;
```

danwalmsley replied:
done

wieslawsoltes commented on src/Avalonia.Visuals/Rendering/SceneGraph/LineBoundsHelper.cs:
```diff
@@ -0,0 +1,72 @@
+
+            if (p.LineCap != PenLineCap.Flat)
+            {
+                var pts = TranslatePointsAlongTangent(p1, p2, angle - 90, p.Thickness / 2);
```
Use radians for `angle - 90` degrees:
```C#
angle - Math.PI/2.0
```

danwalmsley replied:
done

Commit:
Merge branch 'master' into fixes/calculate-bounds-for-drawline

Commit:
do all calculations in radians.

Commit:
fix demo repro

Commit:
add to render demo

Commit:
restore control catalog.

Commit:
whitespace

Commit:
remove unused code.

Commit:
Merge branch 'master' into fixes/calculate-bounds-for-drawline

wieslawsoltes commented:
LGTM tested on Core2D and perf looks on par with master. Also tested line bounds correctness and could not reproduce artifacts.

Commit:
remove pen argument from DrawOperation and BrushDrawOperation

MarchingCube committed:
Merge branch 'master' into fixes/calculate-bounds-for-drawline

Commit:
Merge branch 'master' into fixes/calculate-bounds-for-drawline

grokys committed:
Don't inflate geometry by stroke width twice.

We're already passing the stroke width to `UpdatePathCache` which should include it in its calculations. Here we were applying the stroke width twice.

grokys committed:
Correctly snap dirty rects to device pixels.

Commit:
Merge branch 'master' into fixes/calculate-bounds-for-drawline

grokys committed:
Added failing PixelRect snapping tests.

grokys committed:
Fix PixelRect snapping.

grokys committed:
Merge branch 'fixes/calculate-bounds-for-drawline' of https://github.com/AvaloniaUI/Avalonia into fixes/calculate-bounds-for-drawline

grokys reviewed (changes requested):
Just a few nits.

grokys commented on src/Avalonia.Visuals/Rendering/SceneGraph/LineBoundsHelper.cs:
```diff
@@ -0,0 +1,68 @@
+
+namespace Avalonia.Rendering.SceneGraph
+{
+    public static class LineBoundsHelper
```
Seems it's public in order to be used in RenderDemo. I think making it internal and linking the file into `RenderDemo` would be a better solution than exposing it to the world.

danwalmsley replied:
done.

Commit:
share linebouds helper with render demo. remove hacky test code.

Commit:
use consistent naming in renderdemo

danwalmsley commented on samples/RenderDemo/MainWindow.xaml:
```diff
@@ -44,6 +44,9 @@
       <TabItem Header="GlyphRun">
         <pages:GlyphRunPage/>
       </TabItem>
+      <TabItem Header="LineBoundsDemo">
```
renamed

danwalmsley commented on samples/RenderDemo/MainWindow.xaml:
```diff
@@ -44,6 +44,9 @@
         <pages:GlyphRunPage/>
       </TabItem>
+      <TabItem Header="LineBoundsDemo">
+        <pages:LineBoundsDemo />
```
renamed

danwalmsley commented on samples/RenderDemo/Controls/LineBoundsDemoControl.cs:
```diff
@@ -0,0 +1,46 @@
+            drawingContext.DrawRectangle(boundPen, LineBoundsHelper.CalculateBounds(p1, p2, pen));
+
+            angle += Math.PI / 360;
+            Dispatcher.UIThread.Post(InvalidateVisual, DispatcherPriority.Background);
```
implemented properly.

grokys commented:
I added a few changes to fix the snapping of rectangles to pixel bounds which was causing render artifacts:

Instead of snapping the top-left point and the width/height we need to snap the top-left and bottom-right points to the pixel grid. Snapping the width/height can result in a rectangle that is one pixel too small, if the top-left moved left/up one pixel
